### PR TITLE
Oriented bounding box support

### PR DESCRIPTION
**Pull request by timmydoza**
## 🔗 Related Issues

No related issues to link. Internal ticket: [FOEPD-4586](https://voxel51.atlassian.net/browse/FOEPD-4586)

## 📋 What changes are proposed in this pull request?

result

Adds native support for oriented (rotated) 2D bounding boxes on `fo.Detection` labels via a scalar `rotation` attribute (radians in `[0, 2π)`, applied in pixel space around the box center, positive = clockwise as displayed — matching the direction of `Polyline.from_rotated_box()` and the radian unit of 3D detections).

- **Explore mode (looker)**: detections with a scalar `rotation` render rotated, with rotation-aware hit testing and hover distances. Grid thumbnails and the modal share the same overlay, so both get it.
- **Annotate mode (lighter)**: a selected box gets a CVAT-style rotate handle (Shift snaps to 15°). Rendering (stroke, resize handles, scrim cutout), hit testing, and edge-resize all operate in the box's local frame, so resizing a rotated box pulls the grabbed edge and keeps the opposite edge fixed at any angle. Rotation persists through the annotation engine, participates in undo/redo, and is exposed as a numeric field in the Position panel (committed value, live preview during the gesture, and direct editing).
- **Video annotation**: rotation interpolates between keyframes along the shortest arc (350° → 10° sweeps 20° through zero), matching CVAT's track semantics. Rotation rides through the synthetic-track pipeline so interpolated in-between frames render rotated, and editing a track's last keyframe step-holds rotation over trailing filler frames.
- **Shared math**: both renderers consume one `rotated-box` module in `@fiftyone/utilities` (corner rotation, point-in-rotated-box, local-frame transform, shortest-arc lerp), so explore and annotate cannot drift apart.

Guardrails: detections with instance masks ignore `rotation` (masks are stored relative to the axis-aligned box); a 3D detection's `[x, y, z]` rotation list passes through untouched; boxes that were never rotated never acquire a `rotation` attribute; storage needs no schema changes (`rotation` is a dynamic attribute).

Out of scope, intentionally: the CVAT integration's existing degree-valued `rotation` round-trip (conversion to radians is planned separately), and rotation-awareness in evaluation IoU / `to_patches` / exporters, which remain axis-aligned.

Note: SDK docs for the convention (`Detection` docstring + user guide) are split into a separate docs PR.

## 🧪 How is this patch tested? If it is not, please explain why.

- **Unit tests** (vitest): shared math (corner rotation in pixel space, hit testing, shortest-arc lerp incl. wrap-around); `DetectionOverlay` rotate gesture (enter/rotate/shift-snap/reset), masked-detection guard, rotated containment, and regression tests for local-frame resize (the 180° drag-inversion case and a 90° fixed-opposite-edge case); the linear propagation agent's rotation interpolation and its no-pollution guarantee; the adapter's rotation persistence rules.
- **E2E tests** (Playwright, `e2e-pw`): `annotate-2d-rotate.spec.ts` (seeded rotation reaches the edit form; form edits persist across a fresh browser context; zeroing overwrites the stored scalar; unrotated boxes never acquire the attribute, verified against the DB) and `annotate-video-rotate-interpolation.spec.ts` (a 6.1 → 0.2 rad keyframe pair sweeps the shortest arc through zero, read back per-frame). Both pass locally in prod mode.
- **Manual testing**: rotated boxes verified in explore (grid + modal, image + video), annotate (handle, resize at multiple angles, undo/redo, sidebar editing, autosave round-trip), and video interpolation including keyframe promotion and the wrap-around case.

### Is this a user-facing change that should be mentioned in the release notes?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release
      notes for FiftyOne users.

Added support for oriented (rotated) bounding boxes: 2D detections can carry a scalar `rotation` attribute (radians) that renders in the App, can be edited in annotate mode via a rotate handle, and interpolates between keyframes in video annotation.

### What areas of FiftyOne does this PR affect?

- [x] App: FiftyOne application changes
- [x] Build: Build and test infrastructure changes
- [ ] Core: Core `fiftyone` Python library changes
- [ ] Documentation: FiftyOne documentation changes
- [ ] Other

[FOEPD-4586]: https://voxel51.atlassian.net/browse/FOEPD-4586?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

---
**Enterprise sync PR**: https://github.com/voxel51/fiftyone-teams/pull/4113

## New Features
- Added rotation support for 2D detection boxes, including handles, snapping, rendering, hit testing, resizing, and rotation-aware cursors.
- Added rotation editing with undo and redo support.
- Preserved 2D rotation across labels, video frames, and propagation workflows.
- Added shortest-path rotation interpolation between video keyframes.

## Bug Fixes
- Fixed rotation persistence when resetting a detection’s rotation to zero.
- Improved rotated-box interaction and edge anchoring.
- Preserved existing mask and 3D rotation behavior.

### CODE REVIEW

File: app/packages/annotation/src/engine/surfaces/lighter/adapters.ts [hunk 1]
```diff
@@ -124,7 +124,17 @@ export const detectionAdapter: LighterAdapter = {
...
-    return { ...data, ...maskData, bounding_box: boundingBox };
+    // Rotation channel: the overlay owns the live scalar rotation for 2D
+    // boxes. Persist it when the gesture set one, or when the stored label
+    // already carried a scalar (so rotating back to exactly 0 overwrites the
+    // stale value). A 3D `[x, y, z]` rotation list flows through `data`
+    // untouched — the overlay reports 0 for it. Optional-called: handles are
+    // duck-typed at this boundary (tests supply plain stubs).
+    const rotation = overlay.getRotation?.() ?? 0;
+    const rotationData =
+      rotation !== 0 || typeof data.rotation === "number" ? { rotation } : {};
```
[the remark this answers is not in the document]
Comment (timmydoza) [reply app/packages/annotation/src/engine/surfaces/lighter/adapters.ts]: Fixed in 62048b3c2. The adapter now skips the rotation channel entirely for masked overlays, so the render-suppressed 0 never reaches persistence and the stored scalar flows through `data` untouched. Added a regression test covering exactly this case.

File: app/packages/core/src/components/Modal/Sidebar/Annotate/Edit/Position.tsx [hunk 2]
```diff
@@ -209,8 +235,19 @@ export default function Position({ readOnly = false }: PositionProps) {
...
 
+          // write `rotation` only when it carries signal — a nonzero value, or
+          // zeroing out a stored scalar. Never stamp 0 onto boxes that were
+          // never rotated (and never clobber a 3D `[x, y, z]` list).
+          const storedRotation = engine.getLabel(ref)?.rotation;
+          const rotationData =
+            typeof rotation === "number" &&
+            (rotation !== 0 || typeof storedRotation === "number")
+              ? { rotation }
+              : {};
```
[the remark this answers is not in the document]
Comment (timmydoza) [reply app/packages/core/src/components/Modal/Sidebar/Annotate/Edit/Position.tsx:248]: Fixed in 62048b3c2 with the suggested `!Array.isArray(storedRotation)` guard — a scalar typed into the form can no longer replace a 3D `[x, y, z]` rotation list.

File: app/packages/lighter/src/interaction/InteractionManager.ts [hunk 3]
```diff
@@ -1019,11 +1032,20 @@ export class InteractionManager {
...
-              ? "lighter:overlay-drag-end"
-              : "lighter:overlay-resize-end";
-          this.eventBus.dispatch(type, detail);
+          if (interactionState === "ROTATING") {
+            this.eventBus.dispatch("lighter:overlay-rotate-end", {
+              id: handler.id,
+              overlayId: handler.overlay?.id ?? handler.id,
+              startRotation: startRotation ?? 0,
+              rotation: handler.getRotation?.() ?? 0,
+            });
```
[the remark this answers is not in the document]
Comment (timmydoza) [reply app/packages/lighter/src/interaction/InteractionManager.ts]: Fixed in 62048b3c2. ROTATING now finalizes outside the spatial drag gate (a meaningful angle change can ride on a sub-threshold pointer move, and there's no selection-click ambiguity since the pointer went down on the rotate handle), dispatching only when the angular delta exceeds 1e-4.

File: e2e-pw/src/oss/specs/annotate-2d-rotate.spec.ts [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+test.describe.serial("2D rotated bounding boxes", () => {
+  test("a seeded rotation reaches the edit form's rotation field", async ({
+    modal,
+  }) => {
+    await modal.sidebar.annotate.selectActiveLabel("cat", 0);
+
+    await expect
+      .poll(() => fieldNum(modal, "rotation.rotation"))
+      .toBeCloseTo(SEEDED_ROTATION, 4);
+  });
```
[the remark this answers is not in the document]
Comment (timmydoza) [reply e2e-pw/src/oss/specs/annotate-2d-rotate.spec.ts:128]: Not an issue in practice: `getUniqueDatasetNameWithPrefix` appends a random 7-character suffix at module evaluation, and a serial retry runs in a fresh worker that re-evaluates the module — so a retried `beforeAll` creates a differently-named dataset rather than colliding. This is also the established convention across every spec in the suite; changing the factory's semantics is out of scope for this PR.

**Comment by benjaminpkane** [comment]
Leaving some quick thoughts, take or leave:

1. **CVAT.** I noticed the CVAT importer stores `rotation` in degrees, and with this PR the App would start reading that attribute as radians, so datasets people have already pulled from CVAT might look different after upgrading. You've got the round-trip down as a follow-up, which makes sense. I just wondered whether the `cvat.py` conversion is small enough to ride along here so the two conventions don't overlap in stored data for a release.
2. **Docs.** Makes sense that this rides the same dynamic-attribute path as the 3D `rotation`. For the docs PR, the things I'd find useful next to the existing 3D entry are radians, the sign convention (the `from_rotated_box` docstring says counter-clockwise and this PR says clockwise on screen, both true given y-down, but seeing them together would help), and that rotation is ignored when a mask is present.
3. **e2e specs.** The first three tests in `annotate-2d-rotate.spec.ts` are exactly the pattern I've been moving the suite toward, seeding through the dataset factory and reading back in a fresh context. The fourth could probably do the same instead of going through `executePythonCode`. The annotate specs are already Python-free on the DGVA branch, so that would keep it that way. Two smaller thoughts: armed events tend to be cheaper in CI than `expect.poll`, and a single `toHaveScreenshot` of the rotated box would be a nice way to show it actually renders rotated, since the specs currently only read the form field.

**Comment by timmydoza** [comment]
> Leaving some quick thoughts, take or leave:
> 
> 1. **CVAT.** I noticed the CVAT importer stores `rotation` in degrees, and with this PR the App would start reading that attribute as radians, so datasets people have already pulled from CVAT might look different after upgrading. You've got the round-trip down as a follow-up, which makes sense. I just wondered whether the `cvat.py` conversion is small enough to ride along here so the two conventions don't overlap in stored data for a release.
> 2. **Docs.** Makes sense that this rides the same dynamic-attribute path as the 3D `rotation`. For the docs PR, the things I'd find useful next to the existing 3D entry are radians, the sign convention (the `from_rotated_box` docstring says counter-clockwise and this PR says clockwise on screen, both true given y-down, but seeing them together would help), and that rotation is ignored when a mask is present.
> 3. **e2e specs.** The first three tests in `annotate-2d-rotate.spec.ts` are exactly the pattern I've been moving the suite toward, seeding through the dataset factory and reading back in a fresh context. The fourth could probably do the same instead of going through `executePythonCode`. The annotate specs are already Python-free on the DGVA branch, so that would keep it that way. Two smaller thoughts: armed events tend to be cheaper in CI than `expect.poll`, and a single `toHaveScreenshot` of the rotated box would be a nice way to show it actually renders rotated, since the specs currently only read the form field.

Thanks @benjaminpkane! Awesome stuff. 

1 - Correct, CVAT importer uses degrees, and this PR uses radians. This is deliberate. I discussed it with @ehofesmann and we want it to be consistent with the `rotation` attribute for 3d detections [as described here](https://docs.voxel51.com/user_guide/using_datasets.html#d-detections). We will have to deal with `cvat.py` though, which I'll do in a follow up PR soon, acknowledging that it will only solve future imports and not existing ones. 

2 - Great point, and I'll add this clarification to the docs PR.

3 - Thanks for the tips! Applied and pushed the screenshot and the Python-free rewrite. Kept expect.poll for now since these assert on the form field directly, but happy to swap once the DGVA patterns land.

**Review by benjaminpkane — APPROVED** [review]
